### PR TITLE
Language statistics bar for repositories

### LINKED ISSUE #2165: Languages Statusbar
Are there any plans for gitea to get a status-bar that shows how much of different code its used in the git?

And when you click on it you see the different languages that are used and how many % of the code that is used of that language in the git?

Its a really neat feature that i really miss!

### PR DESCRIPTION
**Pull request by lafriks**
Replaces #4824 after accidentally breaking branch :)

Fixes #2165

Implements language statistic calculation using https://github.com/src-d/enry library.

TODO:
- [x] Function to calculate language stats for git repository
- [x] Save language stats to database
- [x] Display language stats bar ~~(cleanup depends on src-d/enry#233)~~
- [ ] ~~Add support for `linguist-vendored` in `.gitattributes` file~~ (this could be added later in other PR)
- [x] Async update language stats to database when repository changes
- [x] Initial language stats calculation for existing repositories

Screenshots:
image

image

image

### CODE REVIEW

**Comment by kolaente** [comment]
Could the async update be solved via some kind of cron which looks every couple minutes if a repo has been updated in since the last sync and then updates? 
That way you could also get the initial sync "for free". It would also cover the case where a user imports a repo afterwards.

**Comment by lafriks** [comment]
I'm working on a lib for embedded/persistent async job queue currently that could be used here and also for other async tasks

**Comment by lunny** [comment]
Please resolve the conflicts

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by lafriks** [comment]
I still need to finish work on async queue first

**Comment by zeripath** [comment]
Hi @lafriks what about the modules/queue/Queue?

File: models/repo_language_stats.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+type LanguageStatList []*LanguageStat
+
+func (stats LanguageStatList) loadAttributes() {
+	for i := range stats {
+		stats[i].Color = enry.GetColor(stats[i].Language)
+	}
+}
+
+func (repo *Repository) getLanguageStats(e Engine) (LanguageStatList, error) {
+	stats := make(LanguageStatList, 0, 6)
```
Comment (lunny) [inline models/repo_language_stats.go:38]: Maybe you should give the commitID as a parameter?

File: models/repo_list.go [hunk 2]
```diff
@@ -61,6 +64,25 @@ func (repos RepositoryList) loadAttributes(e Engine) error {
 	for i := range repos {
 		repos[i].Owner = users[repos[i].OwnerID]
 	}
+
+	// Load primary language.
+	stats := make(LanguageStatList, 0, len(repos))
```
Comment (lunny) [inline models/repo_list.go:69]: Maybe we need another column `is_default_branch` on LanguageStat.

File: models/repo_language_stats.go (hunk 1 above)
Comment (lafriks) [reply models/repo_language_stats.go:38]: Why commit id? It laods stats from db not git repo

File: models/repo_list.go (hunk 2 above)
Comment (lafriks) [reply models/repo_list.go:69]: Stats will be available only for default branch

**Comment by lafriks** [comment]
@zeripath I needed something more persistent so I implemented it same as code indexer queue

**Comment by lafriks** [comment]
It is now complete and ready for review

**Comment by lunny** [comment]
CI failed.

**Comment by lafriks** [comment]
@lunny fixed

File: modules/indexer/stats/db.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	if err = repo.GetIndexerStatus(models.RepoIndexerTypeStats); err != nil {
+		return err
+	}
+
+	gitRepo, err := git.OpenRepository(repo.RepoPath())
+	if err != nil {
+		return err
+	}
+	// Get latest commit for default branch
```
Comment (lunny) [inline modules/indexer/stats/db.go:33]: `defer gitRepo.Close()`

File: modules/git/repo_language_stats.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+const fileSizeLimit int64 = 16 * 1024 * 1024
+
+// GetLanguageStats calculates language stats for git repository at specified commit
+func (repo *Repository) GetLanguageStats(commitID string) (map[string]float32, error) {
+	r, err := git.PlainOpen(repo.Path)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (lunny) [inline modules/git/repo_language_stats.go:28]: `defer r.Close()`

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
And another nit is the language bar is two narrow to click.

File: modules/indexer/stats/db.go (hunk 3 above)
Comment (lafriks) [reply modules/indexer/stats/db.go:33]: done

File: modules/git/repo_language_stats.go (hunk 4 above)
Comment (lafriks) [reply modules/git/repo_language_stats.go:28]: go-git does not have such method

**Comment by lafriks** [comment]
@lunny made language bar height a bit bigger

**Comment by lunny** [comment]
Could you add an option to disable the feature? For some personal usage, they want to disable it. :(

**Comment by lafriks** [comment]
@lunny what would that give if that is disabled? stats calculation is actually quite fast, for gitea it calculates in about 1 second.

File: models/repo_language_stats.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+			Color:      "#cccccc",
+			Percentage: float32(math.Round(float64(other)*10) / 10),
+		})
+	}
+	return topstats, nil
+}
+
+// UpdateLanguageStats updates the language statistics for repository
+func (repo *Repository) UpdateLanguageStats(commitID string, stats map[string]float32) error {
+	sess := x.NewSession()
```
Comment (guillep2k) [inline models/repo_language_stats.go:79]:
```suggestion
	sess := x.NewSession()
	defer sess.Close()
```
This should certainly help. 😉

File: models/repo_language_stats.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	oldstats, err := repo.getLanguageStats(sess)
```
Comment (guillep2k) [inline models/repo_language_stats.go:85]: ~~Try adding this. Works for me!~~
#### EDIT: Not required; found another way

```suggestion
	// Make sure to lock the repository for this
	if _, err := sess.ID(repo.ID).Cols("`id`").Update(repo); err != nil {
		return err
	}
	oldstats, err := repo.getLanguageStats(sess)
```

File: models/repo_language_stats.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+				IsPrimary:  llang == topLang,
+				Language:   lang,
+				Percentage: perc,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+	// Delete old languages
+	if _, err := sess.Where("`repo_id` = ? AND `commit_id` != ?", repo.ID, commitID).Delete(&LanguageStat{}); err != nil {
```
Comment (guillep2k) [inline models/repo_language_stats.go]: I've finally found the culprit!!!

A DML statement with a non-equality key will lock the entire table. As a workaround, I've placed the following code, although ***not here*** but as early as possible (right after `for lang, perc := range stats {`):

```
	for {
		forDeletion := make([]*LanguageStat,0,20)
		if err := sess.Where("`repo_id` = ? AND `commit_id` != ?", repo.ID, commitID).Limit(100).Find(&forDeletion); err != nil {
			return err
		}
		if len(forDeletion) == 0 {
			break
		}
		ids := make([]int64,len(forDeletion))
		for i := range forDeletion {
			ids[i] = forDeletion[i].ID
		}
		if _, err := sess.In("`id`", ids).Delete(&LanguageStat{}); err != nil {
			return err
		}
	}
```

This will ensure that `DELETE` will only lock the old IDs instead of the whole repository range.

It would _probably_ also work with something like:
```
DELETE FROM `language_stat` WHERE `id` IN ( SELECT `id` FROM `language_state` WHERE `repo_id` = ? AND `commit_id` != ?)
```
But I didn't check.

File: web_src/js/index.js [hunk 8]
```diff
@@ -2,7 +2,7 @@
 /* exported timeAddManual, toggleStopwatch, cancelStopwatch, initHeatmap */
 /* exported toggleDeadlineForm, setDeadline, updateDeadline, deleteDependencyModal, cancelCodeComment, onOAuthLoginClick */
 
-import 'jquery.are-you-sure';
+import 'jquery.are-you-sure/jquery.are-you-sure';
```
Comment (silverwind) [inline web_src/js/index.js]: This seems unnecessary as that file is already the `main` export:

```console
$ jq .main node_modules/jquery.are-you-sure/package.json
"jquery.are-you-sure.js"

File: models/repo_language_stats.go (hunk 5 above)
Comment (lafriks) [reply models/repo_language_stats.go:79]: Thanks! That was it, don't know how I missed that :man_facepalming:

File: web_src/js/index.js (hunk 8 above)
Comment (lafriks) [reply web_src/js/index.js]: My vscode lint was complaining as to if you are importing directory it should have index.js if not full path to js file needs to be specified. But anyway this is out of scope for this PR

Comment (silverwind) [reply web_src/js/index.js]: Sounds like a bug in your linter probably caused by the dot in the module name which is very unusual but valid.

**Comment by lafriks** [comment]
@lunny can you comment on this more?

File: routers/org/home.go [hunk 9]
```diff
@@ -85,7 +85,6 @@ func Home(ctx *context.Context) {
 		OrderBy:            orderBy,
 		Private:            ctx.IsSigned,
 		Actor:              ctx.User,
-		IsProfile:          true,
```
Comment (6543) [inline routers/org/home.go:88]: unrelated change ?

Comment (lafriks) [reply routers/org/home.go:88]: No it's not unrelated, that option was not needed anymore

**Review by lunny — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by lafriks** [comment]
Let's do it :)

**Comment by ALZlper** [comment]
image

Is it possible to hide some languages? Maybe have some settings for it, although I woudl hide SVG by default :)

**Comment by 6543** [comment]
@ALZlper pleace ofen a new Issue, you can reference this pull but dont komment in a merged pull
